### ISSUE TITLE
feat(models): qualify and register Qwen3.8 (Qwen3.8-27B) on the existing qwen3_5 path

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1163]
## Summary

Qualify `mlx-community/Qwen3.8-27B-4bit` as a supported checkpoint, register the Qwen3.8 generation in the docs and detection tables, and turn the config keys mlxcel currently ignores into explicit reads or explicit assertions.

This is **not** a new model family port. Qwen3.8-27B ships `model_type: "qwen3_5"` / `Qwen3_5ForConditionalGeneration` and is architecturally identical to Qwen3.5-27B. It already loads and runs on `main` with zero code changes. The deliverable is qualification, registration, and closing the silent-ignore gaps that make that "works today" accidental rather than guaranteed.

## Research baseline (2026-08-15)

- Repository: `mlx-community/Qwen3.8-27B-4bit`, revision `3e6447f082e89cc7f0bc6e5441afd38dfce760ff`
- License: Apache-2.0, public, ungated. Converted from `Qwen/Qwen3.8-27B` with mlx-vlm 0.6.8
- Pipeline: `image-text-to-text`. 3 shards, 16.05 GB, 2180 tensors
- mlxcel revision under test: `main` @ `6e781fcf7e70a8509c9b4617e6bacb48692a2d5a` (v0.5.1), release binaries built 2026-08-14
- Hardware: Apple Silicon, Metal backend, load average 3.4-4.1 during measurement

Upstream's own framing (Qwen3.8-27B model card): *"Built on the architectural foundation of Qwen3.5."*

### Architectural identity with Qwen3.5-27B

The weight map is identical. Normalizing layer indices and diffing `model.safetensors.index.json` against the local `qwen3.5-27b-4bit`:

```
count35 2180  count38 2180
--- only in 3.5 ---   (none)
--- only in 3.8 ---   (none)
```

Same 64 layers, `hidden_size` 5120, `intermediate_size` 17408, `head_dim` 256, 24 Q heads / 4 KV heads, 48/16 linear value/key heads, `full_attention_interval` 4, vocab 248320, and the same Qwen3-VL vision tower (depth 27, hidden 1152, patch 16, merge 2, `deepstack_visual_indexes: []`).

Upstream agrees: mlx-vlm's `mlx_vlm/models/qwen3_5/config.py` and `vision.py` are byte-identical between the 2026-07-11 tree and current `main` (v0.6.14). The open upstream PR [Blaizzy/mlx-vlm#1899](https://github.com/Blaizzy/mlx-vlm/pull/1899) "Add Qwen3.8-27B support" contains **no architectural change** either: it is FP8 block-dequant, ModelOpt NVFP4/MXFP4 adaptation, and MTP splitting. None of it is needed for the plain `{"group_size": 64, "bits": 4, "mode": "affine"}` mlx-community 4-bit build.

### Config delta vs. Qwen3.5-27B

Only metadata moved. Diff of the two `config.json` files, sorted:

| Key | Qwen3.5-27B | Qwen3.8-27B | mlxcel today |
|---|---|---|---|
| `text_config.output_gate_type` | absent | `"swish"` | ignored |
| `text_config.partial_rotary_factor` | absent (nested only) | `0.25` (also nested) | ignored at top level, read from `rope_parameters` |
| `text_config.mlp_only_layers` | `[]` | absent | `#[serde(default)]` → `[]` |
| `text_config.tie_word_embeddings` | absent | `false` | parsed |
| `language_model_only` | absent | `false` | ignored |
| `generation_config` (nested in config.json) | absent | present | ignored |
| `temperature` / `top_k` / `top_p` / `do_sample` at top level | absent | present | ignored |
| `generation_config.json` `temperature` | 0.6 | 1.0 | read |
| `processor_config.json` `image_processor_type` | `Qwen2VLImageProcessorFast` | `Qwen3VLImageProcessor` | ignored |
| `processor_config.json` sizing | `size.{shortest,longest}_edge` | flat `min_pixels`/`max_pixels` (same values) | `preprocessor_config.json` is byte-identical and is what the loader reads |
| `tokenizer_config.json` `tokenizer_class` | `TokenizersBackend` | `Qwen2Tokenizer` | irrelevant, `tokenizer.json` is used |

`Qwen35Config` (`src/models/qwen3_5.rs:50-105`) carries no `deny_unknown_fields`, so every one of the new keys is silently dropped. The six fields without a serde default (`model_type`, `hidden_size`, `num_hidden_layers`, `num_attention_heads`, `num_key_value_heads`, `vocab_size`) are all present. That is why the checkpoint loads.

## Verification performed (2026-08-15, `main` @ `6e781fcf`)

Everything below was run against the real checkpoint on the release binaries, not synthetically.

**Static preconditions**, each of which could have made the load silently wrong:

- `layer_types` is ignored by mlxcel; the linear/full schedule is computed positionally as `!(i+1).is_multiple_of(full_attention_interval)` (`src/models/qwen3_5.rs:174-176`). Checked all 64 entries of the checkpoint's `layer_types` against that rule: **0 mismatches**.
- `mlp_only_layers` is absent, which would make every layer MoE if `num_experts > 0` (`qwen3_5.rs:178-182`). The checkpoint has no `num_experts`: dense, so the branch is unreachable.
- `PatchEmbed::from_weights` (`src/vision/encoders/qwen3_vl.rs:118-157`) has no quantized branch. The checkpoint's vision tower carries **0** `scales`/`biases` tensors; `vision_tower.patch_embed.proj.{weight,bias}` are plain. Safe.
- `deepstack_visual_indexes` is `[]`, which is the only value upstream accepts for qwen3_5 and the case mlxcel is written for (`src/vision/qwen3_5_vl.rs:35`).

**CLI text.** `mlxcel generate -m models/qwen3.8-27b-4bit -p "What is 17 times 23?" -n 300 --temp 0` → `391`. Correct.

**CLI vision.** 224x224 synthetic PNG (red circle + blue square), prompt "Describe the shapes and their colors" → `Inserted 1 Qwen VL image token blocks (49 total image tokens)`, answer names the red circle correctly.

**Server text.** `POST /v1/chat/completions`, temp 0 → `content: "391"`, `reasoning_content` populated and correctly separated from `content`, `finish_reason: "stop"`, `usage.prompt_tokens: 66`.

**Server vision.** base64 `image_url` data URI + text part → `"The image contains a red circle and a blue square."` Both shapes correct. `usage.prompt_tokens: 102`, which includes the 49 image tokens (no repeat of the Florence-2 undercount in #1086).

**Server multi-turn prompt cache.** Snapshot donation from epic #1148 carries to this checkpoint:

```
turn1: prompt=48  cached=0
turn2: prompt=77  cached=61
turn3: prompt=107 cached=90
```

**Decode throughput**, 3 repeats, same prompt, `--temp 0`, load average 3.8:

| run | tok/s |
|---|---|
| 1 | 18.20 |
| 2 | 18.21 |
| 3 | 18.24 |

Median 18.21 tok/s, spread 0.04. This box carries background load, so treat the absolute number as an order-of-magnitude baseline, not a benchmark record.

### Cache footprint: at the architectural minimum, and the architecture is expensive

Measured by fitting `snapshot_bytes` per entry against prompt length (four points, distinct `prompt_cache_key` per point, 24 GiB snapshot budget so nothing is evicted):

| prompt_tokens | bytes/entry |
|---|---|
| 1,610 | 195,887,104 |
| 6,410 | 514,654,208 |
| 14,410 | 1,034,747,904 |
| 25,610 | 1,772,945,408 |

Fit: `bytes/entry = 91,251,006 + 65,637.6 x tokens`, i.e. **~87 MiB length-independent state plus 64.1 KiB/token**.

The per-token term matches the theoretical f16 minimum for this architecture exactly: 16 full-attention layers x 4 KV heads x 256 head_dim x 2 (K and V) x 2 bytes = **65,536 B/token**. So mlxcel is **not** over-allocating: KV is stored only for the 16 full-attention layers, not for all 64. The 48 gated-delta layers contribute only the length-independent ~87 MiB recurrent + conv state. `paged_blocks_allocated` and `paged_bytes_reserved` are both 0 (this family owns its cache), so the startup line `Paged KV block budget: 2825536 blocks (64 layers, 32-token blocks)` is a cap, not a reservation.

The cost is intrinsic to the checkpoint: `head_dim` 256 is unusually large, so even at a 3:1 linear:full layer ratio the KV is 64 KiB/token. At the native 262,144-token context that is **16 GiB of KV state** for a single sequence, on top of the 15 GiB of weights. Worth stating in the docs.

Two flags that look like mitigations are not. `--kv-bits 8` and `--kv-cache-mode int8` each leave the stored footprint **byte-for-byte identical** to the fp16 default (514,654,208 at 6,410 tokens; 1,772,945,408 at 25,610 tokens under all three settings), and no startup line reports which KV mode is actually active for this family. `qwen3_5` is on the symmetric-Turbo allowlist (`src/lib/mlxcel-core/src/cache/turbo/allowlist.rs:102`), so a Turbo mode may behave differently; that was not measured. Whether the mode is ignored for model-owned caches or the snapshot stores dequantized state is left to the snapshot-capacity follow-up issue.

### 1. The checkpoint is nowhere in the docs or the test set

`docs/supported-models.md` mentions this family in exactly three lines (`:22`, `:96`, `:523`) and names no validated checkpoint. There is no Qwen3.5 or Qwen3.8 paragraph, and no pinned 27B checkpoint outside `tests/tensor_parallel_real_models.rs:410-418`.

### 2. Three config keys are load-bearing upstream and silently ignored here

- **`output_gate_type: "swish"`** originates in vLLM (`qwen_gdn_linear_attn.py`), which selects the gated-delta output-gate activation from `{silu, swish, sigmoid}` and treats `swish` as an alias for `silu`. mlxcel hardcodes SiLU in `RMSNormGated::forward` (`src/models/gated_delta.rs:700-733`), so it is **accidentally correct** for this checkpoint. A future checkpoint declaring `sigmoid` would load and produce silently wrong output. mlx-vlm has the same hole (zero occurrences of the key anywhere upstream).
- **`language_model_only`** is `false` here so nothing changes, but upstream has an open bug ([Blaizzy/mlx-vlm#1812](https://github.com/Blaizzy/mlx-vlm/issues/1812)) where a vision-stripped checkpoint sets it `true` and fails to load; the fix PR #1855 was closed unmerged. If a `language_model_only: true` Qwen3.8 build appears, mlxcel will look for a vision tower that is not there.
- **`rope_parameters.mrope_interleaved: true`** is never read (`InterleavedMRoPE` is hardcoded at `src/models/qwen3_next.rs:556-557`). It matches by construction today.

### 3. Constants that must not be defaulted

mlx-vlm's dataclass defaults disagree with this checkpoint on two token ids, and mlxcel has its own `qwen35_vlm_token_defaults()` (`src/loading/vlm.rs:628-634`):

| Field | mlxcel/mlx-vlm default | Qwen3.8-27B actual |
|---|---|---|
| `vision_start_token_id` | 248045 | **248053** |
| `vision_end_token_id` | 248046 | **248054** |
| `image_token_id` | 248056 | 248056 |
| `video_token_id` | 248057 | 248057 |

The config supplies both explicitly and `qwen_vl_token_ids` (`src/loading/vlm.rs:611-626`) reads `vision_start_token_id`, so the working path is correct. But `vision_end_token_id` is not read at all, and a fallback to 248045 would silently mis-segment vision spans in MRoPE position assignment. Likewise `rope_theta` here is **1e7**, not the 1e5 default, and the checkpoint spells the type key `rope_type` rather than `type`.

### 4. Two upstream fixes landed after mlxcel's reference pin (PREMISE CORRECTED: both were already in the tree)

`references/mlx-vlm` is pinned at `8e2638b` (2026-07-11), which predates this model. Two post-pin commits have a Rust analogue:

- [`24910756`](https://github.com/Blaizzy/mlx-vlm/commit/24910756) (#1741) chunked MRoPE: slice `position_ids` to the current chunk when `position_ids.shape[-1] > seq_length`. Relevant to chunked prefill.
- [`a2482f29`](https://github.com/Blaizzy/mlx-vlm/commit/a2482f29) (#1805) structured output on a padded vocab: the docstring names this exact shape, *"Qwen3.5 pads 248077 real tokens up to 248320 rows"*. A grammar/JSON-schema token mask sized to the real vocab will be wrong against the padded `lm_head`.

The other two post-pin commits (`162e7af8` threadgroup probe, `224f92a4` gated_delta vendoring) are MLX-Python specific with no Rust counterpart.

**Correction, verified before implementation.** Neither fix needed porting. Both behaviors were already present on `main` @ `865bede2`, so the deliverable became "confirm, pin with a regression test, and correct this issue" rather than "port".

- **#1805 (padded-vocab structured-output mask): already correct.** `apply_structured_mask_to_logits` (`src/server/structured.rs:736-788`) already sizes the bias from `vocab_size_hint`, the model's logits axis, and masks every position in `[matcher_vocab, vocab_size_hint)`. Its doc comment at `:718-735` already names this exact case. What landed instead is `apply_mask_covers_the_qwen3_8_padded_lm_head` in `tests/structured_outputs.rs`, which pins the 248077-into-248320 shape at its production magnitudes over a synthetic 248,077-entry byte-level tokenizer, so a refactor that starts sizing the bias from the matcher fails with the checkpoint's own numbers.
- **#1741 (chunked MRoPE position slice): already correct, and stronger.** `Qwen35Model::forward_with_mrope_state` (`src/models/qwen3_5.rs`) already sliced the stored `position_ids` to `[cache_offset, cache_offset + seq_len)` and additionally validated `pos_shape[1] == batch` before reuse, which is a superset of #1741. It was untestable without a checkpoint, so the reuse decision moved into a pure `mrope_position_source` helper covered by four unit tests (chunked walk, overrun, batch mismatch and wrong rank, nothing stored). Behavior is unchanged.

## Implementation plan

1. **Detection and metadata.** Confirm `qwen3_5` + `vision_config` routes to `ModelType::Qwen35VLM` (it does, `src/models/detection.rs:132-141`) and add a config-shape test pinned to this checkpoint's exact `config.json`, so a future serde change that starts rejecting `output_gate_type` or `language_model_only` fails a test rather than a user's load.
2. **Assert the ignored keys instead of dropping them.** Read `output_gate_type` and reject any value other than `swish`/`silu` with a named error; read `mrope_interleaved` and reject `false`; read `language_model_only` and reject `true` until the vision-less path exists. Cheap insurance, and it converts three silent-wrong modes into named failures.
3. **Remove the token-id fallbacks that can be wrong.** Require `vision_start_token_id` from the config for this family and read `vision_end_token_id`; do not default either to the Qwen3.5-era constants.
4. ~~**Port the two upstream fixes**~~ Both were already implemented. Confirm with evidence, pin each with a regression test, and correct the issue (see the correction under "Two upstream fixes" above).
5. **Docs.** Add Qwen3.8 to `docs/supported-models.md` alongside Qwen3.5-VL, name `mlx-community/Qwen3.8-27B-4bit` as the validated checkpoint with the measured numbers, and state the three unsupported surfaces explicitly: no MTP speculative decoding, no video input, no audio (there is no audio tower).
6. **Test set.** Add the checkpoint to the skip-if-missing real-model tests next to the existing `qwen3.5-27b-4bit` entry.

## Explicitly out of scope, tracked separately

Each of these is a property of the `qwen3_5` family or of mlxcel generally, not of Qwen3.8, and each can ship as its own PR:

- **`reasoning_effort` request surface.** Qwen3.8's chat template is the first shipped template in this tree that *validates* its kwargs, which exposes two defects in the server's template handling. Filed separately.
- **MTP speculative decoding.** `Qwen/Qwen3.8-27B` ships 15 `mtp.*` tensors but every mlx-community MLX conversion drops them, and mlx-community publishes the drafter separately as `Qwen3.8-27B-MTP-bf16` (`model_type: qwen3_5_mtp`). mlxcel has no `qwen3_5_mtp` drafter and no `MtpTarget` for this family. Filed separately.
- **Video input.** Upstream qwen3_5 supports video end to end; mlxcel's Qwen VL path is image-only and rejects `--video` with a named error. Filed separately.
- **Snapshot capacity default.** At an agent-scale prompt this model's multi-turn cache hit rate is 0% under the 512 MiB default. Filed separately.

## Non-goals

- **Not omni.** Despite the family naming, Qwen3.8-27B has no audio tower. `Qwen/Qwen3.8-27B/config.json` top-level keys are exactly `architectures, image_token_id, language_model_only, model_type, text_config, tie_word_embeddings, transformers_version, video_token_id, vision_config, vision_end_token_id, vision_start_token_id` — no `audio_config`, `audio_tower`, `thinker_config`, or `talker_config`, and the weight map has only `model.language_model.*`, `model.visual.*`, `mtp.*`, `lm_head.weight`. It is a vision-only VLM (images and video). `modeltype:omni` does not apply.
- **Not the FP8/NVFP4 builds.** `Qwen/Qwen3.8-27B-FP8` and the ModelOpt NVFP4 variants need the block-dequant plumbing from the still-open upstream PR #1899. Out of scope; the plain mlx-community affine 4-bit build is the target.
- **Not Qwen3.8-2.4T-A95B.** The MoE sibling is a separate size class (`modelsize:xlarge`) and would route through `qwen3_5_moe`.

## Acceptance criteria

- [x] `mlx-community/Qwen3.8-27B-4bit` named in `docs/supported-models.md` with the validated surfaces and the three unsupported ones
- [x] Config-shape test pinned to the checkpoint's `config.json`, covering `output_gate_type`, `language_model_only`, `mrope_interleaved`, absent `mlp_only_layers`, and the nested `generation_config`
- [x] `output_gate_type != swish|silu`, `mrope_interleaved: false`, and `language_model_only: true` each fail with a named error rather than loading
- [x] `vision_start_token_id` required from config for the `qwen3_5` family with no fallback to 248045 (`vision_end_token_id` deliberately not read: nothing in the tree consumes it, so adding the field would be dead code). Read from config with no silent fallback to 248045/248046
- [x] ~~#1741 chunked-MRoPE slice and #1805 padded-vocab mask ported with tests~~ Premise corrected: both were already implemented on `main`. Confirmed with file evidence and pinned with regression tests instead
- [x] Real-checkpoint smoke test added next to the existing `qwen3.5-27b-4bit` entries, skip-if-missing
- [x] CLI text and CLI image re-verified against the checkpoint after the change; decode throughput within noise of the 18.2 tok/s baseline

**Comment by inureyes** [comment]
Follow-ups split out of this issue and tracked separately: #1164 (reasoning_effort request surface), #1165 (qwen3_5_mtp drafter), #1166 (Qwen VL video input), #1167 (model-aware snapshot capacity default).
